Optimize TIME_CONVERT/DATE_TIME_CONVERT predicates

Jackie-Jiang opened the pull request:
## Description
Solve #6910 

Add `TimePredicateFilterOptimizer` to optimize TIME_CONVERT/DATE_TIME_CONVERT function with range/equality predicate to directly apply the predicate to the inner expression.
E.g. `dateTimeConvert(col, '1:MILLISECONDS:EPOCH', '1:MILLISECONDS:EPOCH', '30:MINUTES') >= 1620830760000` can be optimized to `col >= 1620831600000`

After optimizing the time convert function to directly work on the time column, the pruner and all the indexes can be applied.

In `HybridQuickstart`, for query `select count(*) from airlineStats where dateTimeConvert(DaysSinceEpoch, '1:DAYS:EPOCH', '1:SECONDS:EPOCH', '30:MINUTES') <= 1388534400`, notice the difference for `numSegmentsProcessed` (segment pruning) and `numEntriesScannedInFilter` (using index)

Original:
```
{
  "resultTable": {
    "dataSchema": {
      "columnNames": [
        "count(*)"
      ],
      "columnDataTypes": [
        "LONG"
      ]
    },
    "rows": [
      [
        289
      ]
    ]
  },
  "exceptions": [],
  "numServersQueried": 2,
  "numServersResponded": 2,
  "numSegmentsQueried": 41,
  "numSegmentsProcessed": 40,
  "numSegmentsMatched": 1,
  "numConsumingSegmentsQueried": 10,
  "numDocsScanned": 289,
  "numEntriesScannedInFilter": 9456,
  "numEntriesScannedPostFilter": 0,
  "numGroupsLimitReached": false,
  "totalDocs": 9810,
  "timeUsedMs": 7,
  "offlineThreadCpuTimeNs": 2874333,
  "realtimeThreadCpuTimeNs": 907709,
  "segmentStatistics": [],
  "traceInfo": {},
  "minConsumingFreshnessTimeMs": 1621752098439
}
```
Optimized:
```
{
  "resultTable": {
    "dataSchema": {
      "columnNames": [
        "count(*)"
      ],
      "columnDataTypes": [
        "LONG"
      ]
    },
    "rows": [
      [
        289
      ]
    ]
  },
  "exceptions": [],
  "numServersQueried": 2,
  "numServersResponded": 2,
  "numSegmentsQueried": 41,
  "numSegmentsProcessed": 1,
  "numSegmentsMatched": 1,
  "numConsumingSegmentsQueried": 10,
  "numDocsScanned": 289,
  "numEntriesScannedInFilter": 0,
  "numEntriesScannedPostFilter": 0,
  "numGroupsLimitReached": false,
  "totalDocs": 9783,
  "timeUsedMs": 2,
  "offlineThreadCpuTimeNs": 83668,
  "realtimeThreadCpuTimeNs": 0,
  "segmentStatistics": [],
  "traceInfo": {},
  "minConsumingFreshnessTimeMs": 1621752181985
}
```

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+    FilterKind filterKind = FilterKind.valueOf(filterFunction.getOperator());
+    List<Expression> operands = filterFunction.getOperands();
+    if (filterKind == FilterKind.AND || filterKind == FilterKind.OR) {
+      operands.replaceAll(this::optimize);
```
when did we start using this coding convention?

sajjad-moradi replied:
why replaceAll? You can do `operands.forEach(this::optimize)` instead. Here you want to apply optimize function on all operands. Although operand.replaceAll does the work, but it's a bit confusing and there's no need to replace the list elements.

Jackie-Jiang replied:
This should be faster because it avoids the per element `get()` and `set()` and the redundant boundary check. We want to avoid the stream API, which might cause performance issue

Jackie-Jiang replied:
@sajjad-moradi For this optimizer we can do that because all the changes are applied in-place. Will add a note and change it

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+      Expression expression = operands.get(0);
+      if (expression.getType() == ExpressionType.FUNCTION) {
+        Function expressionFunction = expression.getFunctionCall();
+        String functionName = StringUtils.remove(expressionFunction.getOperator(), '_');
```
why?

Jackie-Jiang replied:
We ignore the underscore within the function name so that both `dateConvert` and `date_convert` are supported. Currently we don't have a central place to canonicalize the function names, so we have to remove the underscore here.

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+  private void optimizeTimeConvert(Function filterFunction, FilterKind filterKind) {
+    List<Expression> filterOperands = filterFunction.getOperands();
+    List<Expression> timeConvertOperands = filterOperands.get(0).getFunctionCall().getOperands();
+    Preconditions.checkArgument(timeConvertOperands.size() == 3,
```
this should be done somewhere else right?

Jackie-Jiang replied:
Currently the argument check is performed on the server side, but the optimization is performed before routing the query. We can consider moving the argument check to broker to block bad query, but that is out of the scope of this PR

kishoreg commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+ *   </li>
+ * </ul>
+ *
+ * NOTE: This optimizer is followed by the {@link MergeRangeFilterOptimizer}, which can merge the generated ranges.
```
can you add some examples?

Jackie-Jiang replied:
Done

kishoreg reviewed (commented):
Awesome! mostly cosmetic comments.

Can you provide some example queries -  before and after 

Also, show the performance improvement with and without this change

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/QueryOptimizer.java:
```diff
@@ -32,12 +33,14 @@
 public class QueryOptimizer {
   private static final List<FilterOptimizer> FILTER_OPTIMIZERS = Arrays
-      .asList(new FlattenAndOrFilterOptimizer(), new NumericalFilterOptimizer(), new MergeEqInFilterOptimizer(), new MergeRangeFilterOptimizer());
+      .asList(new FlattenAndOrFilterOptimizer(), new MergeEqInFilterOptimizer(), new NumericalFilterOptimizer(),
```
Yes, the order matters. Will add some comments here

amrishlal replied:
We should add comments here to describe dependencies between optimizers.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/QueryOptimizer.java:
```diff
@@ -59,11 +62,13 @@ public void optimize(BrokerRequest brokerRequest, @Nullable Schema schema) {
-    if (filterExpression != null) {
-      for (FilterOptimizer filterOptimizer : FILTER_OPTIMIZERS) {
+    for (FilterOptimizer filterOptimizer : FILTER_OPTIMIZERS) {
+      if (filterExpression != null && filterExpression.getType() == ExpressionType.FUNCTION) {
```
Currently the filter expression can only be function or boolean literal, so the intention here is to short-circuit the extra filter optimizer if the filter is already optimized to a boolean literal.
While this is not closely related to the scope of this PR, and in the future we might want to support column/udf as the filter, so reverting this part to keep it flexible.

amrishlal replied:
+1

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/MergeRangeFilterOptimizer.java:
```diff
@@ -114,26 +112,6 @@ public FilterQueryTree optimize(FilterQueryTree filterQueryTree, @Nullable Schem
     }
   }
 
-  /**
```
Extracted into a separate class so that it can be used by other optimizers

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/utils/Range.java:
```diff
@@ -0,0 +1,119 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.optimizer.filter.utils;
```
This is a helper class for the filter optimizers. Moved it one level up to `org.apache.pinot.core.query.optimizer.filter`

sajjad-moradi commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+    List<Expression> operands = filterFunction.getOperands();
+    if (filterKind == FilterKind.AND || filterKind == FilterKind.OR) {
+      operands.replaceAll(this::optimize);
+    } else if (filterKind.isRange() || filterKind == FilterKind.EQUALS) {
```
How about NOT_EQUALS filter?

Jackie-Jiang replied:
Added some note in the javadoc of the class explaining why it is not supported

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+    List<Expression> operands = filterFunction.getOperands();
+    if (filterKind == FilterKind.AND || filterKind == FilterKind.OR) {
+      operands.replaceAll(this::optimize);
+    } else if (filterKind.isRange() || filterKind == FilterKind.EQUALS) {
```
Are we not rewriting for `NOT_EQUALS, IN, NOT_IN`?

Jackie-Jiang replied:
Added some note in the javadoc of the class explaining why it is not supported

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+      switch (filterKind) {
+        case GREATER_THAN: {
+          // millisToFormat(millis) > n
+          // -> millisToFormat(millis) >= n + 1
+          // -> millis >= formatToMillis(n + 1)
```
Not sure why you need to change the input value here by doing +1. Seems like all that needs to be done here is to properly set lowerInclusive boolean value?

Jackie-Jiang replied:
We have to do `+1` here, or it won't be correct. E.g. `millisToSeconds(millis) > 0` is not equivalent to `millis > 0` because of the flooring of the value. Will add the example into the comment

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java:
```diff
@@ -0,0 +1,330 @@
+      long granularityMillis = new DateTimeGranularitySpec(dateTimeConvertOperands.get(3).getLiteral().getStringValue())
+          .granularityToMillis();
+
+      // Step 1: Convert output range to millis range
```
Can step 1 and step 2 code be refactored and made common between `optimizeTimeConvert `and `optimizeDateTimeConvert` functions?

Jackie-Jiang replied:
I considered using the same code for both `time_convert` and `date_time_convert`, but that would introduce unnecessary overhead for `time_convert` because it does not require SDF and bucketing support. Basically it's a trade-off between performance and code complexity

amrishlal commented on pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizerTest.java:
```diff
@@ -0,0 +1,163 @@
+
+  @Test
+  public void testTimeConvert() {
+    testTimeConvert("timeConvert(col, 'MILLISECONDS', 'MILLISECONDS') > 1620830760000",
+        new Range(1620830760001L, true, null, false));
```
I think this should made to be `new Range(1620830760000L, false, null, false))`

Jackie-Jiang replied:
These 2 ranges are equivalent

amrishlal replied:
Yes, but my point is that we shouldn't be changing input values unless absolutely needed. In this case the +1 can be easily avoided.

amrishlal replied:
Also it seems like, `timconvert(col, 'MILLISECONDS', 'MILLISECONDS') > 9223372036854775807` (long max value) will turn into `col >= 9223372036854775807` which seems incorrect. A test case on this would be useful.

Jackie-Jiang replied:
The algorithm is to handle the general case, so as long as the optimized one is correct we should be good.
For the super large/small value, I'll add a check that the millis is in valid range. If not, the optimizer won't change the filter.

amrishlal replied:
I think it would be much simpler and accurate if you don't do +1. That way it will handle all cases and you won't need to add a if statement for large values.

Jackie-Jiang replied:
We can avoid the `+1` only when the output format is `MILLISECONDS`, and it will complicate the step 2 (convert millis range to input range). Since they are equivalent, I'd prefer just keeping it simple. Added a check that millis must be in valid range to avoid unexpected overflow/underflow.

amrishlal replied:
I see, 2071-1-1 is the current max timestamp. I guess I can live with that. Although it seems like all the semantic checks should be done outside and before the optimizers are run (but that could be refactored at a later time).

amrishlal commented on pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizerTest.java:
```diff
@@ -0,0 +1,163 @@
+    testTimeConvert("TIME_CONVERT(col, 'MILLISECONDS', 'MILLISECONDS') < 1620917160000",
+        new Range(null, false, 1620917160000L, false));
+    testTimeConvert("timeconvert(col, 'MILLISECONDS', 'MILLISECONDS') BETWEEN 1620830760000 AND 1620917160000",
+        new Range(1620830760000L, true, 1620917160001L, false));
```
I think we should avoid changing input values unless absolutely needed. In this case the Range should be `new Range(1620830760000L, true, 162091716000L, true))`

Jackie-Jiang replied:
Special case the same input output time unit case to directly remove the `TIME_CONVERT` transform. Although it should be super rare because it is basically no-op.

Jackie-Jiang commented:
@kishoreg @sajjad-moradi @amrishlal Thanks for the review. Addressed the comments and added example queries in the PR description. For more query examples, check the `TimePredicateFilterOptimizerTest`
